### PR TITLE
NET9 Support

### LINKED ISSUE #4846: Add NET9 target for the framework tests
.NET 9 will be coming out in about a month. It's been RC for a bit. We should start getting the tests running against it. We should consider waiting to merge until .NET 9 hits stable so as to avoid forcing contributors to pull preview bits of IDEs.

### PR DESCRIPTION
**Pull request by Lachstec**
Changes that are needed in order to support .NET 9.0
Fixes #4846 

Work in progress PR in order to prepare for the release of version 9.0.

### CODE REVIEW

**Comment by Lachstec** [comment]
@dotnet-policy-service agree

File: src/NUnitFramework/Directory.Build.props [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <DisableImplicitNuGetFallbackFolder>true</DisableImplicitNuGetFallbackFolder>
     <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
     <NUnitLibraryFrameworks>net462;net6.0</NUnitLibraryFrameworks>
-    <NUnitRuntimeFrameworks>net462;net6.0;net8.0</NUnitRuntimeFrameworks>
+    <NUnitRuntimeFrameworks>net462;net6.0;net9.0</NUnitRuntimeFrameworks>
```
Comment (manfred-brands) [inline src/NUnitFramework/Directory.Build.props:11]: Don't drop Net 8.0, just add 9.0

**Review by manfred-brands — CHANGES_REQUESTED** [review]
Thanks @Lachstec for your contribution

We need to keep net8.0

File: src/NUnitFramework/Directory.Build.props (hunk 1 above)
Comment (Lachstec) [reply src/NUnitFramework/Directory.Build.props]: Done

File: nuget/nunitlite/nunitlite.nuspec [hunk 2]
```diff
@@ -33,7 +33,7 @@ How to use this package:
       <group targetFramework="net6.0">
         <dependency id="NUnit" version="[$version$]" />
       </group>
-	  <group targetFramework="net8.0">
+	  <group targetFramework="net9.0">
```
Comment (manfred-brands) [inline nuget/nunitlite/nunitlite.nuspec:36]: Same here. Keep .net 8.0

File: src/NUnitFramework/Directory.Build.props [hunk 3]
```diff
@@ -25,7 +25,7 @@
 
   <PropertyGroup Condition="'$(BuildingInsideVisualStudio)' != 'true'">
     <!-- Ideally this is always enabled, but that tends to hurt developer productivity -->
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
Comment (manfred-brands) [inline src/NUnitFramework/Directory.Build.props]: No.

Comment (Lachstec) [reply src/NUnitFramework/Directory.Build.props]: Already reverted, accidentally commited it while testing locally

File: nuget/nunitlite/nunitlite.nuspec (hunk 2 above)
Comment (Lachstec) [reply nuget/nunitlite/nunitlite.nuspec:39]: done

**Comment by manfred-brands** [comment]
It looks like the linux build failed due to: `terminate called after throwing an instance of 'PAL_SEHException'`
This is normally a "illegal pointed" which in managed code we should not have.
We will revisit once .NET 9.0 is released.

**Comment by Lachstec** [comment]
I think so to. I spent some time trying to figure out what the issue could be, but it is likely related to some internal things of .NET. Build is failing with the same exception on MacOS.

**Comment by stevenaw** [comment]
I've tried rerunning the builds as RC2 was released this past Tuesday. The present state is all three builds (Windows, Linux, Mac) fail with the following error:

> Error: /home/runner/work/nunit/nunit/src/NUnitFramework/windows-tests/obj/Release/net8.0-windows10.0.19041.0/WinRT.SourceGenerator/Generator.WinRTAotSourceGenerator/WinRTGlobalVtableLookup.g.cs(5,25): error CS0116: A namespace cannot directly contain members such as fields, methods or statements [/home/runner/work/nunit/nunit/src/NUnitFramework/windows-tests/windows-tests.csproj::TargetFramework=net8.0-windows10.0.19041.0]

We'll have to look into this

**Comment by Lachstec** [comment]
I will look into it This Weekend. Thank you for trying with the new RC.

**Comment by stevenaw** [comment]
This may actually be unrelated to .NET 9, and instead related  related to a change in the latest version of the SDK.

There's a workaround listed here: https://github.com/dotnet/sdk/issues/44026

@Lachstec This fix may actually be better as a separate PR so that main can continue to build. We'd welcome a second PR here if you're interested, or I could put one up myself later tonight, in about 11hr

**Comment by Lachstec** [comment]
@stevenaw Thought so, as I already stumbled on this Issue while trying to fix the build. I will open a PR with the fix mentioned there in a minute.

**Comment by Lachstec** [comment]
Tests are passing, so we should be good to go when version 9 drops.

**Review by stevenaw — APPROVED** [review]
Agreed, thanks for keeping an eye on this one @Lachstec 
This looks good to me as well. Thanks for the contribution!

**Comment by stevenaw** [comment]
.NET 9 has launched today 🎉

@Lachstec The PR is listed as a `Draft`. Is it ready for review on your end?
@manfred-brands you'd previously requested some changes here. Does it look alright to you now?

File: global.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "sdk": {
-    "version": "8.0.100",
+    "version": "9.0.100-rc.2",
```
Comment (manfred-brands) [inline global.json:3]:
```suggestion
    "version": "9.0.100",
```

File: src/NUnitFramework/testdata/TestFixtureSourceData.cs [hunk 5]
```diff
@@ -72,10 +72,7 @@ public StaticProperty_SameClass(string arg, string expected) : base(arg, expecte
         {
         }
 
-        public static object[] StaticProperty
-        {
-            get { return new object[] { new object[] { "StaticPropertyInClass" } }; }
-        }
+        public static object[] StaticProperty => new object[] { new object[] { "StaticPropertyInClass" } };
```
Comment (manfred-brands) [inline src/NUnitFramework/testdata/TestFixtureSourceData.cs:75]: Revert this unrelated change as it is not related to .net 9.0 change

**Review by manfred-brands — CHANGES_REQUESTED** [review]
Two changes requested

File: global.json (hunk 4 above)
Comment (Lachstec) [reply global.json]: Has been changed

File: src/NUnitFramework/testdata/TestFixtureSourceData.cs (hunk 5 above)
Comment (Lachstec) [reply src/NUnitFramework/testdata/TestFixtureSourceData.cs]: Reverted it. Sorry for committing in the first place, happened on accident by my lsp

**Comment by Lachstec** [comment]
CI fails because of formatting issues. Should I address them in a separate pr?

**Comment by manfred-brands** [comment]
> CI fails because of formatting issues. Should I address them in a separate pr?

Ok, if the changes to the `StaticProperty` is because of the new net9.0 target, then leave it in (revert the revert)

**Review by manfred-brands — APPROVED** [review]
@Lachstec  Thanks for your patience.

**Comment by Lachstec** [comment]
No problem, hope that it works out now ;)

**Review by mikkelbu — APPROVED** [review]

**Comment by stevenaw** [comment]
LGTM. Thanks for your contribution @Lachstec !
